### PR TITLE
🎉 New source: Partnerstack [low-code CDK]

### PR DESCRIPTION
**Pull request by elliottrabac**
## What
*Describe what the change is solving*
*It helps to add screenshots if it affects the frontend.*

Hacktoberfest: https://github.com/airbytehq/connector-contest/issues/207

## How
*Describe the solution*

## Recommended reading order
1. `x.java`
2. `y.python`

## 🚨 User Impact 🚨
Are there any breaking changes? What is the end result perceived by the user? If yes, please merge this PR with the 🚨🚨 emoji so changelog authors can further highlight this if needed.

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</details>

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub and connector version bumped by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</details>

<details><summary><strong>Connector Generator</strong></summary>

- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed

</details>

## Tests

<details><summary><strong>Unit</strong></summary>

*Put your unit tests output here.*

</details>

<details><summary><strong>Integration</strong></summary>

*Put your integration tests output here.*

</details>

<details><summary><strong>Acceptance</strong></summary>

*Put your acceptance tests output here.*

</details>

### CODE REVIEW

File: airbyte-integrations/connectors/source-partnerstack/source_partnerstack/spec.yaml [hunk 1]
```diff
@@ -0,0 +1,27 @@
+documentationUrl: https://docs.partnerstack.com/reference/auth
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-partnerstack/source_partnerstack/spec.yaml:1]: Change to: `https://docs.airbyte.com/integrations/sources/partnerstack`

File: airbyte-integrations/connectors/source-partnerstack/source_partnerstack/partnerstack.yaml [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+      request_parameters:
+        min_created: "{{ timestamp(config['start_date']) * 1000 }}"
+  retriever:
+    record_selector:
+      $ref: "*ref(definitions.selector)"
+    paginator:
+      type: DefaultPaginator
+      pagination_strategy:
+        type: "CursorPagination"
+        cursor_value: "{{ last_records[-1]['key'] }}"
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-partnerstack/source_partnerstack/partnerstack.yaml:24]: Just curious about the usage of this option.

**Review by marcosmarxm — CHANGES_REQUESTED** [review]
A few comments but the connector looks solid. Please move request my review again when this is ready :)

File: airbyte-integrations/connectors/source-partnerstack/source_partnerstack/partnerstack.yaml (hunk 2 above)
Comment (elliottrabac) [reply airbyte-integrations/connectors/source-partnerstack/source_partnerstack/partnerstack.yaml:24]: Which option are you referring to?

File: airbyte-integrations/connectors/source-partnerstack/source_partnerstack/spec.yaml [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+  $schema: http://json-schema.org/draft-07/schema#
+  title: Partnerstack Spec
+  type: object
+  required:
+    - public_key
+    - private_key
+  additionalProperties: true
+  properties:
+    public_key:
+      type: string
```
[the remark this answers is not in the document]
Comment (elliottrabac) [reply airbyte-integrations/connectors/source-partnerstack/source_partnerstack/spec.yaml:12]: ✅

File: airbyte-integrations/connectors/source-partnerstack/source_partnerstack/spec.yaml (hunk 1 above)
Comment (elliottrabac) [reply airbyte-integrations/connectors/source-partnerstack/source_partnerstack/spec.yaml]: ✅

File: airbyte-integrations/connectors/source-partnerstack/source_partnerstack/partnerstack.yaml (hunk 2 above)
Comment (marcosmarxm) [reply airbyte-integrations/connectors/source-partnerstack/source_partnerstack/partnerstack.yaml:24]: the cursor field, this is the first connector using this feature interpolation

**Comment by marcosmarxm** [comment]
@elliottrabac do you have a test account to share?

**Review by marcosmarxm — CHANGES_REQUESTED** [review]
@elliottrabac can you share the output of `./gradlew airbyte-integrations:connectors:source-partnerstack:integrationTest`?

File: airbyte-integrations/connectors/source-partnerstack/source_partnerstack/partnerstack.yaml (hunk 2 above)
Comment (elliottrabac) [reply airbyte-integrations/connectors/source-partnerstack/source_partnerstack/partnerstack.yaml:24]: Not sure I'm getting your point, but `cursor_value: "{{ last_records[-1]['key'] }}"` is mandatory for the pagination here.

The API returns the object ordered by creation date, and the pagination is possible by adding the last object key in the next request.

Screenshot 2022-11-01 at 03 30 11

**Comment by elliottrabac** [comment]
> @elliottrabac do you have a test account to share?

Unfortunately no :/

I tried all the streams locally and on the VM, working well:
Screenshot 2022-11-01 at 19 44 42

Screenshot 2022-11-01 at 19 44 58

File: airbyte-integrations/connectors/source-partnerstack/source_partnerstack/partnerstack.yaml (hunk 2 above)
Comment (marcosmarxm) [reply airbyte-integrations/connectors/source-partnerstack/source_partnerstack/partnerstack.yaml:24]: @elliottrabac I saw this connector supports incremental syncs. Are you interested in implementing this sync mode for this connector?

**Review by marcosmarxm — CHANGES_REQUESTED** [review]

File: airbyte-integrations/connectors/source-partnerstack/source_partnerstack/partnerstack.yaml (hunk 2 above)
Comment (elliottrabac) [reply airbyte-integrations/connectors/source-partnerstack/source_partnerstack/partnerstack.yaml:24]: Yes sure! Curious to see how it goes with the low code CDK 👍

I'll ping you once done.

**Comment by marcosmarxm** [comment]
@elliottrabac please request my review again when you finish the to solve the comments.

**Comment by marcosmarxm** [comment]
Hello! I'm going to be out of the office this Friday and won't be able to review your contribution again today, I return next Monday. So far, most contributions look solid and are almost done to be approved. As said in [Chris' comment](https://airbytehq.slack.com/archives/C0450PCSDB2/p1667430678866629) all contributions made before 2-November are eligible to receive the prize and have 2 weeks to merge the contributions. But I ensure next week we're going to have your contribution merged. If you have questions about the implementation you can send them in `#hacktoberfest-2022` in Airbyte's Slack.

Sorry the inconvenience and see you again next week, thank you so much for your contribution!

**Comment by elliottrabac** [comment]
Hi @marcosmarxm I tried to implement the incremental sync for this connector but I'm struggling with datetimeformat with the low-code CDK because the format in Partnerstack is epoch timestamp (in ms).

- Do you think there is a workaround for this?
- Do I need to create a custom component for this?

I'm not sure I'll have time to wrap it up in the coming days. As the full refresh connector is working well, do you think we could merge it like that to meet the hacktoberfest deadline and iterate on the incremental sync after?

**Comment by elliottrabac** [comment]
@marcosmarxm I implemented a custom stream slice similar to the [GreenHouse one](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connectors/source-greenhouse/source_greenhouse/greenhouse.yaml).

Acceptance tests with incremental sync:
Screenshot 2022-11-07 at 20 53 20

**Review by marcosmarxm — APPROVED** [review]
Thanks @elliottrabac I'll publish the connector during the week.

**Pull request merged**